Fixed #24179 -- FilteredSelectMultiple widget - add filter field to the right column.

Scalamoosh opened the pull request:
Add functionality to the `SelectFilter` widget to allow searching on the "Selected" side.

The fix is purely for displaying the results, and the value passed to the server/saved to the DB are not affected by what is visible due to search terms. This is to keep it in-line with the "Available" side of the widget.

Implements Ticket [#24179 - FilteredSelectMultiple widget - add filter field to the right column.](https://code.djangoproject.com/ticket/24179)

carltongibson reviewed (commented):
Hi @Scalamoosh — Welcome aboard! 🙂

Thanks for picking this up! 

I think an addition to the admin section of 4.2.txt would be appropriate. 

I think we could do with some tests. `js_tests/admin/SelectFilter2.test.js` has _some_ but they're quite minimal... 🤔 — they check the set up, but no real behaviour. Would you be up for adding some tests for the existing behaviour (add a couple of options, assert they're filtered, and can be moved, ...) in a commit **before** the changes here, and then adding a _can be filtered once selected_ check would be quite minimal. 

(The option to a JS test is a full Selenium test, but I don't suppose that's needed...) 

> The fix is purely for displaying the results, and the value passed to the server/saved to the DB are not affected by what is visible due to search terms. This is to keep it in-line with the "Available" side of the widget.

Just wondering here if we need to make this clear in the help text... 🤔 _Filtering is just for display purposes_ — Don't worry about this for now (likely it's fine as it is...) 

🏅

Scalamoosh commented:
Hi @carltongibson 

I've added some new commits. 

 > I think an addition to the admin section of 4.2.txt would be appropriate. 

Added

> Would you be up for adding some tests for the existing behaviour (add a couple of options, assert they're filtered, and can be moved, ...) in a commit **before** the changes here

Added some tests. Not had much experience ~~using~~ fighting QUnit, so ended up using some API functions rather than testing pure user actions. Let me know if they need updating?

> then adding a _can be filtered once selected_ check would be quite minimal.

Added. Again, I wasn't happy with calling the handlers directly.

felixxm commented:
@Scalamoosh Thanks for this patch :+1: IMO, we still have a big unanswered question about submitting _chosen_ vs. _selected and chosen_ values. We didn't reach a consensus on the ticket (see [comment#3](https://code.djangoproject.com/ticket/24179#comment:3), [comment#4](https://code.djangoproject.com/ticket/24179#comment:4), and [comment#6](https://code.djangoproject.com/ticket/24179#comment:6)), maybe a discussion on the [mailing list](https://code.djangoproject.com/ticket/24179#comment:6) is needed :thinking: Submitting filtered out values can be really confusing for users :exploding_head: (it is for me). We could force users to clear right-side filters before submission :thinking:

carltongibson commented on docs/releases/4.2.txt:
```diff
@@ -35,6 +35,8 @@ Minor features
 * The light or dark color theme of the admin can now be toggled in the UI, as
   well as being set to follow the system setting.
 
+* The "Chosen" column of a ``FilteredSelectMultiple`` widget is now filterable.
```
`filter_horizontal` and `filter_vertical` are both linkable (I think) and are the user-facing API here. I think mentioning that, rather than `FilteredSelectMultiple` is the way to go.

carltongibson commented:
@Scalamoosh Can you have a think about what might be _relatively simple_ to add here...? 

Maybe something like `X selected options not visible`, plus a clear button... 

Or... showing non-matches below but greyed-out... 

Or... 😅

The point being, if we can find a straightforward approach that's sufficient then we can make progress here.

carltongibson commented:
@Scalamoosh assuming it would be simple enough, do you want to give the "Adding a X selected items not shown" message, and we can see what that looks like? Thanks!

Scalamoosh commented:
> @Scalamoosh assuming it would be simple enough, do you want to give the "Adding a X selected items not shown" message, and we can see what that looks like? Thanks!

Something like this? (Cross posted to the mailing list too)

![selected](https://user-images.githubusercontent.com/585291/186267471-06e6443e-5056-4620-8fa9-8c0a80cce1e2.gif)

carltongibson commented:
Hi @Scalamoosh — hat looks good yeah — It's nicely obvious 🙂

The thing that jumps out is, could we add a (muted-text) `click to clear` on the warning banner there? (I would say a link, but why not the why element for a bigger click target. 🤔)

Scalamoosh commented:
Hi @carltongibson - Couldn't find a generic `muted` class that works. 

There is `.quiet` which makes it hardly noticeable on dark mode, and ugly on light. Similar with `.help`, albeit a smaller font-size. 

I have hooked in to the `--breadcrumbs-fg` var, as that has the same background colour. Let me know if you want a specific variable created rather than reusing an unrelated variable?

## `.quiet`

<img width="472" alt="Screenshot 2022-08-24 at 09 14 33" src="https://user-images.githubusercontent.com/585291/186367536-5245d9b5-1947-464f-bd86-975690925c2c.png">
<img width="408" alt="Screenshot 2022-08-24 at 09 16 47" src="https://user-images.githubusercontent.com/585291/186367543-5471ea67-863f-4c15-bea0-d3a6319dd329.png">

## `.help`

<img width="404" alt="Screenshot 2022-08-24 at 09 19 00" src="https://user-images.githubusercontent.com/585291/186367735-1668eab1-a579-42b1-b112-613bc242b6ed.png">
<img width="407" alt="Screenshot 2022-08-24 at 09 19 08" src="https://user-images.githubusercontent.com/585291/186367738-6d200d2b-910f-4fe3-8dd2-7419de6a544d.png">

## `var(--breadcrumbs-fg)`

<img width="407" alt="Screenshot 2022-08-24 at 09 33 37" src="https://user-images.githubusercontent.com/585291/186371181-4f0a3469-54dc-4113-96a1-25aff7d56128.png">
<img width="400" alt="Screenshot 2022-08-24 at 09 33 47" src="https://user-images.githubusercontent.com/585291/186371185-f1ba8a34-2be4-40f9-a24e-4af57d03910f.png">

carltongibson commented:
OK, great. Push your best option, and I'll take a look. 
Thanks!

carltongibson commented:
Hey @Scalamoosh — I think that looks pretty good. I'll ask Mariusz to take a look over the next few days and see if he's in agreement. 

Assuming so... 🤞

Meanwhile are you able to rebase and squash down to a single commit?
(Then a final nick-pic, and we should be there.)  

Thanks for the effort! 🎁

Scalamoosh commented:
Apologies @carltongibson - A messed up git rebase has closed the PR. Can it be re-opened, or does it need a fresh one creating? The branch is still there, with a single commit now. https://github.com/django/django/compare/main...Scalamoosh:django:fix-24179

carltongibson commented:
Open a fresh one. No problem (these things happen)
